networkx/algorithms/centrality/domirank.py

```diff
@@ -0,0 +1,301 @@
+            "The system is large!!! Consider using `analytical = False` function argument for reduced computational time cost."
+        )
+    # Here we create another warning for alpha being supercharged
+    if alpha > 1:
+        if analytical == False:
+            raise nx.NetworkXUnfeasible(
+                "supercharging the competition parameter (`alpha > 1`) requires the `analytical = True` flag."
+            )
+        else:
+            import warnings
+
+            warnings.warn(
+                "You are supercharging the competition in the system by having `alpha > 1`, which is only permitted for the analytical solution!"
+            )
+    if alpha < 0:
+        raise nx.NetworkXUnfeasible(
+            "the competition parameter alpha must be positive - `alpha > 0`."
+        )
+
+    # Here we renormalize alpha with the smallest eigenvalue (most negative eigenvalue) by calling the "hidden" function _find_smallest_eigenvalue()
+    # Note, this function always uses the iterative definition
+    sigma = np.abs(
+        alpha
+        / _find_smallest_eigenvalue(
+            GAdj,
+            max_depth=max_iter // 5,
+            dt=dt,
+            epsilon=epsilon,
+            max_iter=max_iter,
+            patience=patience,
+        )
+    )
+    if analytical:
+        converged = None
+        Psi = sp.sparse.linalg.spsolve(
+            sigma * GAdj + sp.sparse.identity(GAdj.shape[0]), sigma * GAdj.sum(axis=-1)
+        )
+    else:
+        Psi, sigma, converged = _domirank_iterative(
+            GAdj,
+            sigma=sigma,
+            dt=dt,
+            epsilon=epsilon,
+            max_iter=max_iter,
+            patience=patience,
+        )
+    Psi = dict(zip(G, (Psi).tolist()))
+    return Psi, sigma, converged
+
+
+##### THE FUNCTIONS HEREUNDER SHOULD BE HIDDEN FUNCTIONS #####
+def _find_smallest_eigenvalue(
+    G,
+    min_val=0,
+    max_val=1,
+    max_depth=100,
+    dt=0.1,
+    epsilon=1e-5,
+    max_iter=100,
+    patience=10,
+):
+    """
+    This function is simply used to find the smallest eigenvalue, by seeing when the DomiRank algorithm diverges. It uses
+    a kind of binary search algorithm, however, with a bias to larger eigenvalues, as finding divergence is faster than
+    verifying convergence.
+    This function outputs the smallest eigenvalue - i.e. most negative eigenvalue.
+    """
+    import numpy
+    import scipy
+
+    x = (min_val + max_val) / G.sum(axis=-1).max()
+    for __ in range(max_depth):
+        if max_val - min_val < epsilon:
+            break
+        ___, ____, converged = _domirank_iterative(
+            G, sigma=x, dt=dt, epsilon=epsilon, max_iter=max_iter, patience=patience
+        )
+        if converged:
+            min_val = x
+            x = (min_val + max_val) / 2
+        else:
+            max_val = (x + max_val) / 2
+            x = (min_val + max_val) / 2
+    final_val = (max_val + min_val) / 2
+    return -1 / final_val
```

Comment:
Help me understand: this is returning `-1/lambda`, and we are dividing by it in the main function - are we supposed to divide twice?